src/java/org/apache/cassandra/db/rows/ColumnData.java

```diff
@@ -35,6 +40,176 @@
+        {
+            if (!(existing instanceof ComplexColumnData))
+            {
+                Cell<?> existingCell = (Cell) existing, updateCell = (Cell) update;
+
+                Cell<?> result = Cells.reconcile(existingCell, updateCell);
+
+                return modifier.merge(existingCell, result);
+            }
+            else
+            {
+                ComplexColumnData existingComplex = (ComplexColumnData) existing;
+                ComplexColumnData updateComplex = (ComplexColumnData) update;
+
+                DeletionTime existingDeletion = existingComplex.complexDeletion();
+                DeletionTime updateDeletion = updateComplex.complexDeletion();
+                DeletionTime maxComplexDeletion = existingDeletion.supersedes(updateDeletion) ? existingDeletion : updateDeletion;
+
+                Object[] existingTree = existingComplex.tree();
+                Object[] updateTree = updateComplex.tree();
+
+                Object[] cells;
+
+                try (Reconciler reconciler = reconciler(modifier, maxComplexDeletion))
+                {
+                    if (!maxComplexDeletion.isLive())
+                    {
+                        if (maxComplexDeletion == existingDeletion)
+                        {
+                            updateTree = BTree.transformAndFilter(updateTree, reconciler::retain);
+                        }
+                        else
+                        {
+                            existingTree = BTree.transformAndFilter(existingTree, reconciler::retain);
+                        }
+                    }
+                    cells = BTree.update(existingTree, updateTree, existingComplex.column.cellComparator(), (UpdateFunction) reconciler);
+                }
+                return new ComplexColumnData(existingComplex.column, cells, maxComplexDeletion);
+            }
+        }
+
+        @Override
+        public void onAllocatedOnHeap(long heapSize)
+        {
+            modifier.onAllocatedOnHeap(heapSize);
+        }
+
+        @Override
+        public ColumnData insert(ColumnData insert)
+        {
+            return modifier.insert(insert);
+        }
+
+        @Override
+        public ColumnData retain(ColumnData existing)
+        {
+            if (!(existing instanceof ComplexColumnData))
+            {
+                if (activeDeletion.deletes((Cell) existing))
+                {
+                    modifier.delete(existing);
+                    return null;
+                }
+            }
+            else
+            {
+                ComplexColumnData existingComplex = (ComplexColumnData) existing;
+
+                DeletionTime existingDeletion = existingComplex.complexDeletion();
+
+                Object[] existingTree = existingComplex.tree();
+
+                if (existingDeletion.supersedes(activeDeletion))
+                {
+                    try (Reconciler reconciler = reconciler(modifier, existingDeletion))
```

Comment:
If the `existingDeletion` supersedes the `activeDeletion` then surely we have nothing to do here, assuming the complex data is correct?